sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/batch/BulkExecutorTemp.java

```diff
@@ -0,0 +1,1138 @@
+                            "Flushing PKRange {} (batch size: {}) due to FlushItemOperation, Context: {} {}",
+                            thresholds.getPartitionKeyRangeId(),
+                            currentMicroBatchSizeSnapshot,
+                            this.operationContextText,
+                            getThreadInfo());
+
+                        firstRecordTimeStamp.set(-1);
+                        currentMicroBatchSize.set(0);
+                        currentTotalSerializedLength.set(0);
+                        pastIdAndPK.clear();
+
+                        return true;
+                    }
+
+                    // avoid counting flush operations for the micro batch size calculation
+                    return false;
+                }
+
+                CosmosItemIdentity cosmosItemIdentity = BulkExecutorUtil.createCosmosItemIdentity(itemOperation);
+                boolean isItemAbsent = pastIdAndPK.add(cosmosItemIdentity);
+                firstRecordTimeStamp.compareAndSet(-1, timestamp);
+                long age = timestamp - firstRecordTimeStamp.get();
+                long batchSize = currentMicroBatchSize.incrementAndGet();
+                int totalSerializedLength = this.calculateTotalSerializedLength(currentTotalSerializedLength, itemOperation);
+
+                if (batchSize >= thresholds.getTargetMicroBatchSizeSnapshot() ||
+                    age >= this.maxMicroBatchIntervalInMs || totalSerializedLength >= this.maxMicroBatchPayloadSizeInBytes || (preserveOrdering && !isItemAbsent)) {
+                    if (preserveOrdering && !isItemAbsent) {
+                        logger.debug(
+                            "BufferUntil - Flushing PKRange {} due to isItemAlreadyPresent ({}), BatchSize ({}), payload size ({}) or age ({}) " +
+                                "Triggering {}, Context: {} {}",
+                            thresholds.getPartitionKeyRangeId(),
+                            true,
+                            batchSize,
+                            totalSerializedLength,
+                            age,
+                            getItemOperationDiagnostics(itemOperation),
+                            this.operationContextText,
+                            getThreadInfo());
+                    } else {
+                        logger.debug(
+                            "BufferUntil - Flushing PKRange {} due to BatchSize ({}), payload size ({}) or age ({}), " +
+                                "Triggering {}, Context: {} {}",
+                            thresholds.getPartitionKeyRangeId(),
+                            batchSize,
+                            totalSerializedLength,
+                            age,
+                            getItemOperationDiagnostics(itemOperation),
+                            this.operationContextText,
+                            getThreadInfo());
+                    }
+                    firstRecordTimeStamp.set(-1);
```

Comment:
Why do we have a residual setting on `firstRecordTime`, `currentMicroBatchSize`, `pasIdAndPK`, `currentTotalSerializedLength`? The bufferUntil is about to be done at this point.

Reply:
To filter by partition key and id the batches, I have to encounter the duplicate itemOp first and send the latest one in the next batch. I also changed the bufferUntil to cut before the last element. Therefore, for keeping track of the `firstRecordTime`, `currentMicroBatchSize`, `pasIdAndPK`, `currentTotalSerializedLength` I have to add the last item's information to these.